### PR TITLE
✨ Enforce go import order via gci linter

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

 

**What this PR does / why we need it**:
The GCI linter is essentially an improved (imho) version of goimports, which is also compatible with goimports, i.e. gci results in an import format which is compatible to what goimports enforces. (just a bit more strict)

With the configuration in this PR it sorts the imports in this order: (we could merge the second and third block if we want)
```go
<go-sdk-imports>

<external-imports>

<imports from the current repo>
```

I think this would help reviewers and contributors by just enforcing the right import order this way. gci supports `make lint-fix` so nobody has to sort the imports manually after e.g. Intellij added them arbitrarily.

The current situation where we are not linting it automatically but manually on PRs does not seem like an ideal situation and requires additional effort from reviewers and contributors.

P.S. The changes in this PR are a result of the changes in the golang-ci config + running `make lint-fix`.

xref: https://github.com/kubernetes-sigs/cluster-api/pull/5585#discussion_r743537641

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: api/v1alpha3/suite_test.go [hunk 1]
```diff
@@ -22,9 +22,9 @@ import (
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/client-go/kubernetes/scheme"
-	"sigs.k8s.io/cluster-api/internal/envtest"
 	ctrl "sigs.k8s.io/controller-runtime"
-	// +kubebuilder:scaffold:imports
```
Comment (killianmuldoon) [inline api/v1alpha3/suite_test.go:27]: I'm not familiar with kubebuilder - what was this tag doing?

File: bootstrap/kubeadm/api/v1alpha3/suite_test.go [hunk 2]
```diff
@@ -22,9 +22,9 @@ import (
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/client-go/kubernetes/scheme"
-	"sigs.k8s.io/cluster-api/internal/envtest"
 	ctrl "sigs.k8s.io/controller-runtime"
-	// +kubebuilder:scaffold:imports
```
Comment (killianmuldoon) [inline bootstrap/kubeadm/api/v1alpha3/suite_test.go:27]: As above.

File: bootstrap/kubeadm/main.go [hunk 3]
```diff
@@ -43,10 +47,6 @@ import (
 	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
 	"sigs.k8s.io/cluster-api/feature"
 	"sigs.k8s.io/cluster-api/version"
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
-	// +kubebuilder:scaffold:imports
```
Comment (killianmuldoon) [inline bootstrap/kubeadm/main.go:49]: As above.

**Review by killianmuldoon — APPROVED** [review]
/lgtm

+/- a question on kubebuilder tags - are they just artefacts from the creation process?

**Comment by killianmuldoon** [comment]
Thanks for this! Definitely going to be a big help on PRs 😄

File: api/v1alpha3/suite_test.go (hunk 1 above)
Comment (sbueringer) [reply api/v1alpha3/suite_test.go:27]: kubebuidler will add new imports if you generate new APIs or stuff at those tags.

I'll have to take a closer look, but it's usually possible to place them somwhere where they belong and are not moved by `gci`

Comment (sbueringer) [reply api/v1alpha3/suite_test.go:27]: I restored all `// +kubebuilder:scaffold:imports`. I'm not sure it makes a lot of sense to have them though.

I don't think we currently use the kubebuilder CLI at all to create new API types or reconcilers. And if we would use it I'm not sure how valid the result would be (compared to just copy/paste/modify existing API types or reconcilers)

File: .golangci.yml [hunk 4]
```diff
@@ -42,6 +43,8 @@ linters-settings:
   ifshort:
     # Maximum length of variable declaration measured in number of characters, after which linter won't suggest using short syntax.
     max-decl-chars: 50
+  gci:
+    local-prefixes: sigs.k8s.io/cluster-api
```
Comment (vincepri) [inline .golangci.yml:47]: Can we use no local prefixes? Most folks out there with goimports won't necessarily make use of it and we can reduce some friction here

Comment (sbueringer) [reply .golangci.yml]: Done

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**